### PR TITLE
Limit number of concurrent branches

### LINKED ISSUE #1265: Limit number of concurrent branches
A useful debugging option is to limit the number of concurrent branches in a proof. Basically, we are limiting the size of the frontier set. This is extremely important when debugging loop invariants.

---

- [ ] `--breadth LIMIT` option (opposite `--depth`)

### PR DESCRIPTION
**Pull request by andreiburdusa**
---
Fixes #1265 
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/app/exec/Main.hs [hunk 1]
```diff
@@ -247,7 +247,8 @@ data KoreExecOptions = KoreExecOptions
     , smtTimeOut          :: !SMT.TimeOut
     , smtPrelude          :: !(Maybe FilePath)
     , smtSolver           :: !Solver
-    , stepLimit           :: !(Limit Natural)
+    , breadthLimit          :: !(Limit Natural)
```
Comment (ttuegel) [inline kore/app/exec/Main.hs]:
```suggestion
    , breadthLimit        :: !(Limit Natural)
```

File: kore/app/exec/Main.hs [hunk 2]
```diff
@@ -329,6 +332,12 @@ parseKoreExecOptions =
             , ("any-heating-cooling", heatingCooling anyRewrite)
             , ("all-heating-cooling", heatingCooling allRewrites)
             ]
+    breadth =
+        option auto
+            (  metavar "BREADTH"
+            <> long "breadth"
+            <> help "Execute up to BREADTH steps."
```
Comment (ttuegel) [inline kore/app/exec/Main.hs]:
```suggestion
            <> help "Allow up to BREADTH parallel execution branches."
```

File: kore/src/Kore/Exec.hs [hunk 3]
```diff
@@ -220,20 +221,21 @@ execGetExitCode
...
     -> smt ExitCode
-execGetExitCode indexedModule strategy' finalTerm =
+execGetExitCode breadthlimit indexedModule strategy' finalTerm =
     case resolveInternalSymbol indexedModule $ noLocationId "LblgetExitCode" of
         Nothing -> return ExitSuccess
         Just mkExitCodeSymbol -> do
             exitCodePattern <-
                 -- TODO (thomas.tuegel): Run in original execution context.
-                exec indexedModule strategy'
+                exec breadthlimit indexedModule strategy'
```
Comment (ttuegel) [inline kore/src/Kore/Exec.hs]: The `breadthLimit` here should be hardcoded to `1`. Execution must not branch when we are choosing the exit code.

File: kore/src/Kore/Step/Strategy.hs [hunk 4]
```diff
@@ -414,6 +421,8 @@ constructExecutionGraph transit instrs0 searchOrder0 config0 = do
       | []              <- instrs = unfoldWorker rest searchOrder
       | instr : instrs' <- instrs
       = Profile.executionQueueLength (Seq.length rest) $ do
+        when ((not . withinLimit depthLimit . fromIntegral . Seq.length) rest)
```
Comment (ttuegel) [inline kore/src/Kore/Step/Strategy.hs]: Let's factor out this error into a helper function. We should also `unparse` the branches for the error message.

File: kore/test/Test/Kore/Step/Strategy.hs [hunk 5]
```diff
@@ -232,8 +232,8 @@ test_Or =
         ]
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Strategy.hs]:
```suggestion
```

File: kore/src/Kore/Step/Strategy.hs [hunk 6]
```diff
@@ -368,6 +376,11 @@ emptyExecutionGraph config =
 -}
 data GraphSearchOrder = BreadthFirst | DepthFirst deriving Eq
 
+data LimitExceeded = LimitExceeded | Other
```
Comment (ttuegel) [inline kore/src/Kore/Step/Strategy.hs]: What is `Other` for? It would also be nice if the exception type captured the execution queue when it is thrown.

Comment (andreiburdusa) [reply kore/src/Kore/Step/Strategy.hs]: I know I did an ugly thing, but I did this in order to distinguish between the failing of the function and the failing of the test caused by the success of the function. I mostly pushed this to save my work. It was the only way I was able to make it work in almost 6 hours. I'll try to come with a better solution

Comment (andreiburdusa) [reply kore/src/Kore/Step/Strategy.hs]: I think I already have a better idea :))

Comment (ttuegel) [reply kore/src/Kore/Step/Strategy.hs]: It looks like you figured this out below, but you can `catch` or `handle` the `LimitExceeded` exception and use `assertFailure` if the exception is not thrown.

File: kore/src/Kore/Step.hs [hunk 7]
```diff
@@ -66,6 +66,7 @@ import Kore.Unparser
 {- | A strategy primitive: a rewrite rule or builtin simplification step.
  -}
 data Prim rewrite = Simplify | Rewrite !rewrite
+        deriving (Show)
```
Comment (ttuegel) [inline kore/src/Kore/Step.hs]:
```suggestion
    deriving (Show)
```

File: kore/test/Test/Kore/Exec.hs [hunk 8]
```diff
@@ -113,6 +124,78 @@ test_search =
...
+
+    assertion searchType =
+        shouldExeedBreadthLimit searchType `catch`
+            \(_ :: LimitExceeded (Strategy (Prim Rewrite))) -> pure ()
+
+    shouldExeedBreadthLimit :: SearchType -> IO ()
+    shouldExeedBreadthLimit searchType = do
+        a <- actual searchType
+        when (a == expected searchType)
+            $ assertFailure "Did not exceed bredth limit"
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Exec.hs]:
```suggestion
            $ assertFailure "Did not exceed breadth limit"
```

**Pull request merged**